src/test/java/org/junitpioneer/jupiter/CartesianProductTestExtensionTests.java

```diff
@@ -0,0 +1,185 @@
+		}
+
+		@CartesianProductTest({ "0", "1" })
+		void threeBits(String a, String b, String c) {
+			int value = Integer.parseUnsignedInt(a + b + c, 2);
+			Assertions.assertThat(value).isBetween(0b000, 0b111);
+		}
+
+		@CartesianProductTest
+		@DisplayName("S ⨯ T ⨯ U")
+		void nFold(String string, Class<?> type, TimeUnit unit, TestInfo info) {
+			Assertions.assertThat(string).endsWith("a");
+			Assertions.assertThat(type).isInterface();
+			Assertions.assertThat(unit.name()).endsWith("S");
+			Assertions.assertThat(info.getTags()).isEmpty();
+		}
+
+		@CartesianProductTest(factory = "supplyValues")
+		@DisplayName("Can have an explicit factory method over the implicit one")
+		void explicitFactory(String string, TimeUnit unit) {
+			Assertions.assertThat(string).isIn("War", "Peace");
+			Assertions.assertThat(unit.name()).endsWith("S");
+		}
+
+	}
+
+	static CartesianProductTest.Sets nFold() {
+		return new CartesianProductTest.Sets()
+				.add("Alpha", "Omega")
+				.add(Runnable.class, Comparable.class, TestInfo.class)
+				.add(TimeUnit.DAYS, TimeUnit.HOURS);
+	}
+
+	static CartesianProductTest.Sets supplyValues() {
+		return new CartesianProductTest.Sets().add("War", "Peace").add(TimeUnit.SECONDS, TimeUnit.DAYS);
+	}
+
+	@Nested
+	class BadConfigurationTests {
+
+		@Test
+		@DisplayName("Test fails when there is no factory method")
+		void throwsForMissingFactoryMethod() {
+			ExecutionResults results = PioneerTestKit
+					.executeTestMethodWithParameterTypes(BadConfigurationTest.class, "noFactory", int.class);
+
+			assertThat(results)
+					.hasSingleFailedContainer()
+					.withExceptionInstanceOf(AssertionError.class)
+					.hasMessageContaining("not found");
+		}
+
+		@Test
+		@DisplayName("Test fails when the factory is not static")
+		void throwsForNonStaticFactoryMethod() {
+			ExecutionResults results = PioneerTestKit
+					.executeTestMethodWithParameterTypes(BadConfigurationTest.class, "nonStaticFactory", int.class);
+
+			assertThat(results)
+					.hasSingleFailedContainer()
+					.withExceptionInstanceOf(AssertionError.class)
+					.hasMessageContaining("must be static");
+		}
+
+		@Test
+		@DisplayName("Test fails when the factory does not return Sets")
+		void throwsForWrongReturnValueFactoryMethod() {
+			ExecutionResults results = PioneerTestKit
+					.executeTestMethodWithParameterTypes(BadConfigurationTest.class, "wrongReturnFactory", int.class);
+
+			assertThat(results)
+					.hasSingleFailedContainer()
+					.withExceptionInstanceOf(AssertionError.class)
+					.hasMessageContaining("must return");
+		}
+
+		@Test
+		@DisplayName("Test fails if the factory does not produce enough parameters")
+		void throwsForTooFewFactoryMethodParameters() {
+			ExecutionResults results = PioneerTestKit
+					.executeTestMethodWithParameterTypes(BadConfigurationTest.class, "incompleteFactory", int.class,
+						String.class, TimeUnit.class);
+
+			assertThat(results).hasNumberOfDynamicallyRegisteredTests(3).hasNumberOfFailedTests(3);
```

Comment:
Can a message be provided (and asserted) in such case?

Reply:
Even though all tests fail, the container is considered successful. No clue why.

Reply:
There should be a discussion about extending Pioneer assertions as need for it arises, (such as in this case).